### ISSUE TITLE
Evolve Associate into a learned capability router / mixture of specialists

### ISSUE THREAD

**Issue by OriNachum** [agentculture/associate#12]
## Idea

Evolve `associate` from **one non-coding worker backed by one model** into the AgentCulture layer that **associates an action/context with the most appropriate available model, specialist adapter, and toolchain**.

The name becomes literal:

> **Associate maps intent and actions to the most appropriate available capability.**

This should remain a narrow responsibility. Associate answers **“who should do this step?”**; it should not become a workflow/process orchestrator.

- **Culture Nodes** owns process/workflow and determines *what needs to happen*.
- **Associate** selects *which capability/model/specialist should perform the step*.
- **Harness** executes that capability.
- **Lobes** exposes/serves the available models.
- **Colleague / nvsh / robots / other embodiments** consume Associate as a routing layer.

## Why

Today Associate is framed as a fast, reliable worker for read/summarize/find tasks, implemented as an opinionated Pi extension and backed by the `associate` Nemotron lobe. That is useful, but it makes the project identity largely synonymous with one model/persona.

A stronger identity is:

```text
context / action
      |
      v
  Associate
      |
      +--> select capability
      |
      +--> select model / adapter
      |
      +--> select harness/toolchain policy
      |
      v
 execute specialist
      |
      v
    handback
```

The existing read/find/summarize worker can remain the default/general Associate capability and fallback while the routing layer is added incrementally.

## Mixture of specialists

The initial target is a **model-level sparse MoE / mixture of specialists** rather than a classical token-level transformer MoE.

Example on a Jetson AGX Orin:

```text
                        +-- LFM CUDA specialist
                        +-- LFM Docker specialist
request --> LFM router -+-- LFM Linux specialist
                        +-- LFM Python specialist
                        +-- LFM log specialist
                        +-- LFM command-fix specialist
                        +-- generic Associate fallback
```

A particularly attractive implementation is **one LFM2.5 base + many LoRA/adapters**, rather than many full checkpoints:

```text
LFM2.5 base
  |
  +-- router adapter
  +-- cuda adapter
  +-- containers adapter
  +-- linux adapter
  +-- python adapter
  +-- logs adapter
  +-- verifier adapter
```

The router itself can be a small fine-tuned model whose only learned task is:

```text
(context, available specialists) -> specialist selection
```

It should normally route **top-1** for latency/efficiency, with optional top-k / verifier / larger-model escalation when confidence is low.

This makes Orin useful as a **stash of learned skills / hands**, rather than trying to make it compete with larger machines as another general-purpose brain.

## Dynamic specialist registry

Do not hardcode every specialist into the router weights.

Each specialist should register a small capability manifest, for example:

```yaml
name: cuda-debugger
version: 1

when_to_use_me:
  - CUDA runtime errors
  - GPU / unified-memory failures
  - NVIDIA driver/runtime incompatibility

do_not_use_me:
  - generic package-manager failures

model:
  base: lfm2.5-2.6b
  adapter: cuda-debugger

tools:
  - inspect_gpu
  - inspect_memory

output_schema: associate.handback.v1
```

Potential metadata:

- `name`
- `description`
- `when_to_use_me`
- `do_not_use_me`
- base model / adapter
- supported harnesses
- tools / permissions
- output schema
- latency / cost class
- context limit
- hardware requirements
- version
- evaluation score / confidence calibration

Associate should present the currently available registry to the learned router. That gives us:

> **learned routing ability + dynamic capability discovery**

A newly installed specialist should become immediately routable from its manifest; the router can later be fine-tuned again from accumulated routing examples.

## Associate CLI shape

Possible future surface (names illustrative):

```bash
associate specialist list
associate specialist show cuda-debugger
associate specialist install ./cuda-debugger
associate specialist remove cuda-debugger

associate route --json '<task/context>'
associate run '<task>'

associate association create cuda-debugger --dataset routing.jsonl
associate association eval cuda-debugger
associate association export cuda-debugger
```

`associate route --json` should make the decision observable, e.g.:

```json
{
  "specialist": "cuda-debugger",
  "confidence": 0.96,
  "alternatives": [
    {"specialist": "linux-debugger", "confidence": 0.03}
  ],
  "reason": "CUDA OOM / unified-memory pressure"
}
```

The routing record is valuable evaluation/training data; it should be possible to persist it without leaking task content by default.

## Use `agentculture/unsloth-cli` to create new associations

Use [`agentculture/unsloth-cli`](https://github.com/agentculture/unsloth-cli) as the standard AgentCulture training/evaluation/export path for learned associations.

Conceptually, creating a new specialist should be a reproducible pipeline:

```text
behavior/routing examples
        |
        v
   sloth train
        |
        v
    sloth eval
        |
        v
   sloth export
        |
        v
register adapter + manifest in Associate
```

`unsloth-cli` already has the useful lifecycle we want:

```bash
sloth train --config run.toml
sloth eval --adapter DIR --suite suite.jsonl
sloth export --adapter DIR --output OUT
```

Associate should wrap or generate the artifacts/configuration needed for that lifecycle rather than implementing another trainer.

Example:

```bash
associate association init cuda-debugger
# produces manifest + dataset/eval skeleton

associate association train cuda-debugger
# delegates to sloth train

associate association eval cuda-debugger
# delegates to sloth eval

associate association publish cuda-debugger
# sloth export + register adapter/manifest
```

The exact verbs can be decided later; the architectural requirement is that **training is externalized to unsloth-cli and Associate owns the capability contract + routing/evaluation semantics**.

### LFM2.5 support note

`unsloth-cli` is currently proven primarily on Qwen adapter-class models. If LFM2.5 is not yet a supported/validated target, do not special-case training inside Associate. Add/extend LFM2.5 support in `agentculture/unsloth-cli`, then consume that capability here.

This keeps one fine-tuning stack for AgentCulture.

## Training data: associations, not generic instruction tuning

For the router, the important dataset is not “answer this task.” It is:

```json
{
  "context": "...",
  "available_specialists": ["cuda", "containers", "linux", "generic"],
  "expected": "cuda"
}
```

Useful signals can come from real runs:

```text
context
 -> router selection
 -> specialist output
 -> accepted/rejected handback
 -> verifier result
 -> downstream outcome
```

That lets us train routing from **observed usefulness**, not only manually labelled categories.

For specialist adapters, train stable behavior/reflexes rather than changing facts. Mutable state remains in context/RAG/tools.

Examples of good specialist-finetune targets:

- classify/triage an NVIDIA failure
- identify relevant evidence in logs
- select required inspectors/tools
- produce a constrained diagnosis schema
- summarize a code/file/module for another agent
- inspect a diff for a narrow risk class
- extract structured facts
- generate a proposed command without executing it
- verify whether a previous action solved the condition

## nvsh as an early proving ground

`agentculture/nvsh` is a particularly clean first consumer because its failure context is bounded and structured:

```text
command
exit code
PIPESTATUS
stdout/stderr slice
platform facts
cwd
request kind
```

That maps naturally to specialists such as:

- `nvsh-triage`
- `nvsh-cuda`
- `nvsh-memory`
- `nvsh-containers`
- `nvsh-package`
- `nvsh-log-reader`
- `nvsh-fixer`
- `nvsh-verifier`

Important boundary: **the specialist never owns authority or execution**. `nvsh` keeps deterministic policy, approvals, allowlists, execution and auditing. Associate supplies intelligence and handback only.

This also yields unusually clean training episodes:

```text
failure
 -> specialist selected
 -> diagnosis
 -> proposed fix
 -> user approved/rejected
 -> retry result
 -> verification
```

## Relationship to Colleague

This also sharpens the Colleague/Associate split:

- **Colleague**: coherent coding/task agent; handles the task it has been given.
- **Associate**: learned capability router + non-coding specialist lane; answers which available specialist should handle a scoped step.

Colleague can call Associate when it needs a reader, seeker, summarizer, verifier, log analyst, domain expert, etc., without having to know the serving topology or checkpoint names.

The caller should ask for a capability, not a checkpoint.

## Keep routing separate from orchestration

Critical non-goal: do **not** turn Associate into Culture Nodes.

Associate should not own plans, waves, DAGs or long-running workflow state.

The clean boundary is:

```text
Culture Nodes: what happens next?
Associate:      who/what should do this step?
Harness:        execute the selected capability
Lobes:          where/how is the model served?
```

This separation is the main architectural value of the proposal.

## Escalation / fallback

Specialists should fail closed to a more general capability.

Possible policy:

```text
small router
   |
   +-- confidence high --> specialist adapter
   |
   +-- confidence medium --> specialist + verifier
   |
   +-- confidence low --> generic Associate / Nemotron
                              |
                              +--> optional Qwen/Spark escalation
```

The goal is not “small models everywhere.” The goal is **use the cheapest/smallest learned capability that is reliable for this step, while preserving a path to stronger models**.

## Evaluation

Routing needs its own behavioral suite. Measure at least:

- top-1 routing accuracy
- confusion matrix by specialist
- abstain/escalate quality
- latency to routing decision
- end-to-end task success after routing
- regression against generic Associate baseline
- specialist handback schema validity
- wrong-specialist harm / recovery rate
- router behavior when specialists are added/removed dynamically

Every specialist should ship with its own eval suite and minimum acceptance threshold.

Fine-tuning should be justified by a measurable behavioral delta; if prompt-only behavior already meets the threshold, keep the specialist prompt-only.

### Phase 1 — capability contract

- [ ] Define `SpecialistManifest` / association schema.
- [ ] Add registry + discovery.
- [ ] Add deterministic/prompt-only router baseline.
- [ ] Preserve current Nemotron read/find/summarize Associate as fallback.
- [ ] Add `associate route --json` with observable decisions.

### Phase 2 — learned router

- [ ] Define routing dataset schema.
- [ ] Add eval corpus and confusion-matrix reporting.
- [ ] Fine-tune a small router through `unsloth-cli`.
- [ ] Compare learned router vs prompt-only baseline.
- [ ] Add confidence/abstain/escalation policy.

### Phase 3 — LFM2.5 specialist adapters

- [ ] Validate LFM2.5 adapter training/export/serving path.
- [ ] If needed, add LFM2.5 support upstream in `agentculture/unsloth-cli`.
- [ ] Train 2–3 narrow specialists from real tasks.
- [ ] Serve base + adapters on Orin.
- [ ] Benchmark adapter switching, concurrency, memory and TTFT.

### Phase 4 — consumers

- [ ] Integrate `nvsh` as first structured specialist consumer.
- [ ] Let Colleague request Associate capabilities without knowing model ids.
- [ ] Make Lobes expose capability/model availability needed by the registry.
- [ ] Feed accepted/rejected/outcome signals back into routing eval/training data.

## Success criteria

This is successful when:

1. A caller can submit a scoped action/context without naming a checkpoint.
2. Associate selects an available specialist from a dynamic registry.
3. The selected specialist is measurably better/cheaper/faster than always using the generic Associate for its target class.
4. Adding a new specialist does not require changing caller code.
5. New associations can be trained/evaluated/exported reproducibly through `unsloth-cli`.
6. Low-confidence routing reliably falls back/escalates rather than confidently choosing the wrong specialist.
7. Workflow ownership remains outside Associate.
8. Existing Associate behavior remains available as the general fallback during migration.

## Working definition

> **Associate is the learned capability router for AgentCulture: it associates a scoped action and its context with the most appropriate available model, specialist adapter and toolchain, executes that capability through its harness boundary, and hands the result back.**

That makes “associate” a system responsibility rather than the name of one Nemotron persona/checkpoint.
